### PR TITLE
[4.0] Per User accessibility settings

### PR DESCRIPTION
**Pull request by brianteeman**
Although this isnt strictly necessary as a user can control this themselves on their browser if they want to it is generally considered a good idea to offer the ability to increase font, highlight links, high contrast and mono filters.

This PR adds the ability to set all of these at the user level in the user profile. This way they are permanently set for the user and are completely independent of their browser and cookies.

As the new template has an accessibility settings link on the top right user menu I added the link there as well. NOTE: J4 has lost the ability to link directly to a tab - reported elsewhere - so when thats fixed the link will need updating.

It is also easy this way to add further settings in the future if so desired.

image

### CODE REVIEW

File: administrator/templates/atum/index.php [hunk 1]
```diff
@@ -32,6 +32,12 @@
 $hiddenMenu = $app->input->get('hidemainmenu');
 $joomlaLogo = $this->baseurl . '/templates/' . $this->template . '/images/logo.svg';
 
+// Getting user accessibility settings
+$a11y_mono 		= (bool) $app->getIdentity()->getParam('a11y_mono', '');
```
Comment (zero-24) [inline administrator/templates/atum/index.php]:
```suggestion
$a11y_mono      = (bool) $app->getIdentity()->getParam('a11y_mono', '');
```

File: administrator/templates/atum/index.php [hunk 2]
```diff
@@ -32,6 +32,12 @@
...
+$a11y_contrast  = (bool) $app->getIdentity()->getParam('a11y_contrast', '');
```
Comment (zero-24) [inline administrator/templates/atum/index.php:37]:
```suggestion
$a11y_contrast  = (bool) $app->getIdentity()->getParam('a11y_contrast', '');
```

File: administrator/templates/atum/index.php [hunk 3]
```diff
@@ -32,6 +32,12 @@
...
+$a11y_highlight = (bool) $app->getIdentity()->getParam('a11y_highlight', '');
+$a11y_font 		= (bool) $app->getIdentity()->getParam('a11y_font', '');
```
Comment (zero-24) [inline administrator/templates/atum/index.php]:
```suggestion
$a11y_font         = (bool) $app->getIdentity()->getParam('a11y_font', '');
```

File: administrator/templates/atum/index.php [hunk 4]
```diff
@@ -80,16 +86,17 @@
 
 HTMLHelper::getServiceRegistry()->register('atum', 'JHtmlAtum');
 HTMLHelper::_('atum.rootcolors', $this->params);
+
+
```
Comment (zero-24) [inline administrator/templates/atum/index.php]: IMO we can remove one of them ;)

File: administrator/modules/mod_user/tmpl/default.php [hunk 5]
```diff
@@ -39,8 +39,8 @@
 			</a>
 		</div>
 		<div class="dropdown-item">
-			<?php // TODO: route to accessibility settings ?>
-			<a href="#">
+			<?php $route = 'index.php?option=com_users&task=user.edit&id=' . $user->id . '&return=' . base64_encode($uri); ?>
+			<a href="<?php echo Route::_($route); ?>">
```
Comment (zero-24) [inline administrator/modules/mod_user/tmpl/default.php:43]:
```suggestion
			<a href="<?php echo Route::_('index.php?option=com_users&task=user.edit&id=' . $user->id . '&return=' . base64_encode($uri)); ?>">
```

File: administrator/modules/mod_user/tmpl/default.php [hunk 6]
```diff
@@ -39,8 +39,8 @@
 			</a>
 		</div>
 		<div class="dropdown-item">
-			<?php // TODO: route to accessibility settings ?>
-			<a href="#">
+			<?php $route = 'index.php?option=com_users&task=user.edit&id=' . $user->id . '&return=' . base64_encode($uri); ?>
```
Comment (zero-24) [inline administrator/modules/mod_user/tmpl/default.php:42]: When the next line is accepted this line can be removed. ;)

File: administrator/modules/mod_user/tmpl/default.php (hunk 5 above)
Comment (brianteeman) [reply administrator/modules/mod_user/tmpl/default.php:43]: I was replicating the way the other links in this module are done

File: administrator/templates/atum/index.php (hunk 2 above)
Comment (brianteeman) [reply administrator/templates/atum/index.php:37]: What's the difference. Can't see any on my phone
In

**Comment by ciar4n** [comment]
Is there a particular usage for the 'high contrast' fields? If for color blindness, would it be better to set a palette for each color blindness type (dropdown with normal, dueteranopia, protanopia etc.) Considering we now have custom properties for colors, this is something that would be relatively easy.

File: administrator/templates/atum/index.php [hunk 7]
```diff
@@ -32,6 +32,12 @@
 $hiddenMenu = $app->input->get('hidemainmenu');
 $joomlaLogo = $this->baseurl . '/templates/' . $this->template . '/images/logo.svg';
 
+// Getting user accessibility settings
+$a11y_mono      = (bool) $app->getIdentity()->getParam('a11y_mono', '');
+$a11y_contrast  = (bool) $app->getIdentity()->getParam('a11y_contrast', '');
+$a11y_highlight = (bool) $app->getIdentity()->getParam('a11y_highlight', '');
+$a11y_font         = (bool) $app->getIdentity()->getParam('a11y_font', '');
```
Comment (Quy) [inline administrator/templates/atum/index.php]:
```suggestion
$a11y_font      = (bool) $app->getIdentity()->getParam('a11y_font', '');
```

**Comment by brianteeman** [comment]
> Is there a particular usage for the 'high contrast' fields?

Just a fairly standard setting found in these type of tools

> If for color blindness, would it be better to set a palette for each color blindness type (dropdown with normal, dueteranopia, protanopia etc.) 

Thats the nice thing about this code approach - it is easy to enhance it - in a later PR.

**Comment by brianteeman** [comment]
@ciar4n re the colour blindness. I have searched and all I can find are filters to simulate the effect of protanopia  etc which isnt what you want.

**Comment by ciar4n** [comment]
@brianteeman Don't use filters. Color CSS variables can be redefined with a class name. 

Eg.

```css
.color-protanopia {
  --atum-bg: #5f7e92;
  etc. etc.
}
```

Apply the above class to the body for such users requiring these colors. These will replace CSS variables defined under :root

**Comment by brianteeman** [comment]
Sorry I wasn't clear. I do not believe you can make a palette suitable for the issue. Thats one of the reasons we have the contrast rules

**Comment by ciar4n** [comment]
> I do not believe you can make a palette suitable for the issue.

I can't see why not. If protanopia makes it difficult to distinguish between yellows and greens then we create a palette that avoids using both of these colors.

If increasing the contrast makes it easier for those with color blindness (very questionable), then we can assume they would simply adjust their monitor rather than messing with settings on a per-site basis.

Finally, currently this PR doesn't increase the color contrast even thou it is labeled that it does. It inverts the color which is very different.

**Comment by brianteeman** [comment]
>  then we can assume they would simply adjust their monitor rather than messing with settings on a per-site basis.

Personally I agree with you on this and the same applies to all the settings here. But its a fairly standard thing for this type of function and the a11y team said it was useful to have this functionality on a per site basis as well as at the browser/os level

> then we create a palette that avoids using both of these colors.

If you can do that then go for it. Either as a separate Pr or as a pr to this branch

File: administrator/templates/atum/index.php [hunk 8]
```diff
@@ -80,16 +86,16 @@
...
-<html lang="<?php echo $this->language; ?>" dir="<?php echo $this->direction; ?>">
+<html lang="<?php echo $this->language; ?>" dir="<?php echo $this->direction; ?>" <?php echo ($a11y_font ? 'class="a11y_font"' : ''); ?>">
 <head>
 	<jdoc:include type="metas" />
 	<jdoc:include type="styles" />
 </head>
 
-<body class="admin <?php echo $option . ' view-' . $view . ' layout-' . $layout . ($task ? ' task-' . $task : '') . ($monochrome ? ' monochrome' : ''); ?>">
-
+<body class="admin <?php echo $option . ' view-' . $view . ' layout-' . $layout . ($task ? ' task-' . $task : '') . ($monochrome ? ' monochrome' : '') . ($a11y_mono ? ' monochrome' : '') . ($a11y_contrast ? ' a11y_contrast' : '') . ($a11y_highlight ? ' a11y_highlight' : '')?>">
```
Comment (Quy) [inline administrator/templates/atum/index.php]:
```suggestion
<body class="admin <?php echo $option . ' view-' . $view . ' layout-' . $layout . ($task ? ' task-' . $task : '') . ($monochrome ? ' monochrome' : '') . ($a11y_mono ? ' monochrome' : '') . ($a11y_contrast ? ' a11y_contrast' : '') . ($a11y_highlight ? ' a11y_highlight' : ''); ?>">
```

File: administrator/templates/atum/index.php [hunk 9]
```diff
@@ -80,16 +86,16 @@
 
 HTMLHelper::getServiceRegistry()->register('atum', 'JHtmlAtum');
 HTMLHelper::_('atum.rootcolors', $this->params);
+
 ?>
 <!DOCTYPE html>
-<html lang="<?php echo $this->language; ?>" dir="<?php echo $this->direction; ?>">
+<html lang="<?php echo $this->language; ?>" dir="<?php echo $this->direction; ?>" <?php echo ($a11y_font ? 'class="a11y_font"' : ''); ?>">
```
Comment (Quy) [inline administrator/templates/atum/index.php:92]:
```suggestion
<html lang="<?php echo $this->language; ?>" dir="<?php echo $this->direction; ?>" <?php echo ($a11y_font ? 'class="a11y_font"' : ''); ?>>
```

File: administrator/components/com_users/forms/user.xml [hunk 10]
```diff
@@ -195,6 +195,49 @@
...
+		<fieldset name="accessibility" label="COM_USERS_A11Y_SETTINGS_FIELDSET_LABEL">
+			<field
+				name="a11y_mono"
+				type="radio"
+				label="COM_USERS_A11Y_SETTINGS_FIELD_MONOCHROME"
+				class="switcher"
+				default="0"
+				>
+				<option value="0">JNO</option>
+				<option value="monochrome">JYES</option>
```
Comment (Quy) [inline administrator/components/com_users/forms/user.xml:208]: Why not use value `1`? This value is cast as a boolean in the code.

File: administrator/templates/atum/index.php [hunk 11]
```diff
@@ -80,16 +86,16 @@
...
 <head>
 	<jdoc:include type="metas" />
 	<jdoc:include type="styles" />
 </head>
 
-<body class="admin <?php echo $option . ' view-' . $view . ' layout-' . $layout . ($task ? ' task-' . $task : '') . ($monochrome ? ' monochrome' : ''); ?>">
-
+<body class="admin <?php echo $option . ' view-' . $view . ' layout-' . $layout . ($task ? ' task-' . $task : '') . ($monochrome ? ' monochrome' : '') . ($a11y_mono ? ' monochrome' : '') . ($a11y_contrast ? ' a11y_contrast' : '') . ($a11y_highlight ? ' a11y_highlight' : ''); ?>">
```
Comment (Quy) [inline administrator/templates/atum/index.php:98]: There are 2 settings for monochrome.

`($monochrome ? ' monochrome' : '') . ($a11y_mono ? ' monochrome' : '')`

**Comment by ciar4n** [comment]
If the a11y team can define suitable pallettes for each, I will create the pr to apply them.

**Comment by brianteeman** [comment]
@ciar4n there is no team

**Comment by brianteeman** [comment]
@ciar4n I have found out how to get suitable colour palettes.

**Comment by ciar4n** [comment]
I'll get a base started in a PR.

Just seen the number of color variables we are defining. A little excessive so I would suggest we reduce these where possible. Probably best in a separate PR made first.

image

**Comment by brianteeman** [comment]
I would suggest only changing the colours that could have been changed in the template style. That gives us two sets

`:root {--atum-bg-light: #e8e8e8;--toolbar-bg: #f5f5f5;--atum-text-dark: #495057;--atum-text-light: #ffffff;--atum-link-color: #1757a1;--atum-link-hover-color: #0a2747;--atum-special-color: #001b4c;}`

and from the hue 
`:root {--atum-bg-dark: #6b501a;--atum-contrast: #b52c2c;--atum-bg-dark-0: #f7f7f7;--atum-bg-dark-5: #e8e8e8;--atum-bg-dark-10: #d9d9d9;--atum-bg-dark-20: #bababa;--atum-bg-dark-30: #9f9a93;--atum-bg-dark-40: #917e5f;--atum-bg-dark-50: #866832;--atum-bg-dark-70: #524214;--atum-bg-dark-80: #39300e;--atum-bg-dark-90: #252009;}`

We only need to provide three options for the main color barrier types
And so far in the tests a lot of the current colours dont change

**Comment by ciar4n** [comment]
My point is that a lot of the variables are duplicates. Currently we have the following so we need to redefine every variable for each color barrier...

```
--blue: #1a466b;
--primary: #1a466b;
--atum-bg-dark: #1a466b;
```

Better the following so we only redefine `--blue`.

```
--blue: #1a466b;
--primary: var(--blue);
--atum-bg-dark: var(--blue);
```

**Comment by brianteeman** [comment]
If you can give me a list then I can calculate the values

**Comment by ciar4n** [comment]
Hard to say. Colors are been defined quite randomly across the template. Some colors are still hardcoded. Some by color name (--blue, --red etc.). Some by state color (--success, --warning etc.), Also a collection of unnecessary (imo) variations of the same color (--atum-bg-dark-20, --atum-bg-dark-30 etc.).

Presuming all these variables are been used, then I would say all of them.

**Comment by ciar4n** [comment]
Regarding the hardcoded color, we obviously can not redefine them. This would include elements I would expect should change color depending on the color blindness type (eg. buttons).

**Comment by brianteeman** [comment]
That's really why I suggested https://github.com/joomla/joomla-cms/pull/26194#issuecomment-534921350

**Comment by ciar4n** [comment]
Yes but most of them colors are for aesthetics only. As they already follow WCAG AA, they should not really be an issue regarding color blindness. To truly benefit color blindness we should be redefining non-aesthetic colors (button color etc.).

**Comment by brianteeman** [comment]
well I will have to leave that up to you to work out. The tool I found was https://coolors.co/

You input the color and then click on the eye icon and select the color-blindness simulator you want. It will then change the color accordingly. BUT it doesn't change the color value that is displayed :( You have to use a color picker to grab that - unless you can find the js calculation script they are using

**Comment by ciar4n** [comment]
Neat. I use coolors.co but never noticed that option.

**Comment by brianteeman** [comment]
it's certainly a hidden feature

**Comment by Schmidie64** [comment]
I have tested this item :white_check_mark: successfully on 222137da145197958c48f2c1f9f6427176a62251<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26194">issues.joomla.org/tracker/joomla-cms/26194</a>.</sub>

**Comment by shraddhaSankpal27** [comment]
I have tested this item :white_check_mark: successfully on 222137da145197958c48f2c1f9f6427176a62251<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26194">issues.joomla.org/tracker/joomla-cms/26194</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26194">issues.joomla.org/tracker/joomla-cms/26194</a>.</sub>

**Comment by Renuka-S** [comment]
I have tested this item :white_check_mark: successfully on 222137da145197958c48f2c1f9f6427176a62251<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26194">issues.joomla.org/tracker/joomla-cms/26194</a>.</sub>

**Comment by wilsonge** [comment]
Thanks! @ciar4n still interested in your colour PR but looks like it's going to need another round of tweaks for the a11y issues mentioned there first

**Comment by brianteeman** [comment]
Thanks

**Pull request merged**